Refactor tooling management

paulfantom opened the pull request:
just a few things:
- move to go 1.14
- remove `generate-in-docker` target
- use golang jsonnet 
- add `jsonnet-fmt` target to makefile
- remove unused entries in .travis.yml
- do not install tooling binaries system-wide, but only locally
- use separate `go.mod` for tooling binaries
- make `vendor` target a phony one to simplify usage

/cc @coreos/team-monitoring

dgrisonnet commented on .travis.yml:
```diff
@@ -1,8 +1,8 @@
+- "1.14.x"
 go_import_path: github.com/coreos/prometheus-operator
 env:
   - GO111MODULE=on
```
I don't think we need that anymore in 1.14

dgrisonnet commented on Makefile:
```diff
@@ -162,6 +155,7 @@ $(CRD_JSONNET_FILES): $(GOJSONTOYAML_BINARY) $(CRD_YAML_FILES)
 	cat example/prometheus-operator-crd/monitoring.coreos.com_podmonitors.yaml     | gojsontoyaml -yamltojson > jsonnet/prometheus-operator/podmonitor-crd.libsonnet
```
Here we should use ```$(GOJSONTOYAML_BINARY)``` instead of gojsontoyaml.

paulfantom replied:
Command was removed during rebased :shrug:

pgier commented on Makefile:
```diff
@@ -31,26 +24,28 @@ CRD_JSONNET_FILES += jsonnet/prometheus-operator/prometheusrule-crd.libsonnet
 CRD_JSONNET_FILES += jsonnet/prometheus-operator/servicemonitor-crd.libsonnet
 CRD_JSONNET_FILES += jsonnet/prometheus-operator/thanosruler-crd.libsonnet
 
+TOOLS_BIN_DIR ?= $(shell pwd)/tmp/bin
```
Nice idea!  I will be happy to not worry about failures due to different versions of the tools installed to go/bin.

paulfantom commented:
- Rebased
- incuded tidying of `scripts/go.{mod,sum}` files in `vendor` target
- removed formatting *.libsonnet files
- added `-mod=readonly -modfile=scripts/go.mod` to tooling installation to clarify which `go.mod` file is used and to discard `vendor/` directory as it is not relevant for tools.

paulfantom commented on Makefile:
```diff
@@ -182,12 +176,17 @@ $(TO_BE_EXTENDED_DOCS): $(EMBEDMD_BINARY) $(shell find example) bundle.yaml
+.PHONY: jsonnet-fmt
+jsonnet-fmt: $(JSONNETFMT_BINARY)
+	# *.*sonnet will match *.jsonnet and *.libsonnet files but nothing else in this repository
+	find . -name *.jsonnet -not -path "*/vendor/*" -print0 | xargs -0 jsonnetfmt -i
```
note to self: 
```suggestion
	find . -name *.jsonnet -not -path "*/vendor/*" -print0 | xargs -0 $(JSONNETFMT_BINARY) -i
```

paulfantom commented:
Rebased to fix conflicts

paulfantom commented:
Rebased to fix conflicts

paulfantom commented:
Rebased to fix conflicts and amend usage of jsonnetfmt binary.

paulfantom commented:
Fix downloading jsonnet binaries.

paulfantom commented:
Rebased.

paulfantom commented:
Seems like it works now. Travis is red due to quay.io outage.

simonpasquier commented on Makefile:
```diff
@@ -132,16 +129,15 @@ image: .hack-operator-image .hack-prometheus-config-reloader-image
 ##############
 
+.PHONY: vendor
 vendor:
```
maybe change the target name since it's not about vendoring only anymore.

paulfantom replied:
Separated non-vendoring commands into a separate target.

dgrisonnet commented on Makefile:
```diff
@@ -15,36 +15,33 @@ REPO_PROMETHEUS_OPERATOR_LINT?=quay.io/coreos/prometheus-operator-lint
+PO_DOCGEN_BINARY:=$(TOOLS_BIN_DIR)/po-docgen
+CONTROLLER_GEN_BINARY := $(TOOLS_BIN_DIR)/controller-gen
+EMBEDMD_BINARY=$(TOOLS_BIN_DIR)/embedmd
+GOBINDATA_BINARY=$(TOOLS_BIN_DIR)/go-bindata
```
I don't think we are using `go-bindata` anymore. It was removed along with CRD management

dgrisonnet commented on scripts/generate/build-rbac-prometheus-operator.sh:
```diff
@@ -6,10 +6,12 @@ set -o pipefail
+
+rm -rf "$TMP"
+mkdir "$TMP"
+jsonnet -J scripts/generate/vendor scripts/generate/prometheus-operator-rbac.jsonnet > "$TMP/po.json"
```
Instead of storing the json data in a temporary file can't we use the same mechanism, we have in kube-prometheus? https://github.com/coreos/kube-prometheus/blob/master/build.sh#L18-L21

paulfantom replied:
I did it a bit differently to make it a bit clearer and easier to understand. Also no `xargs` as it isn't very portable.

paulfantom commented:
> I don't see the magical removal of extra files generated for folks on non linux machines :) But I could be missing it?

Is it a problem here also? I know it was a problem in kube-prometheus, but not here :thinking:

dgrisonnet commented:
We don't have this problem here because the `tmp` directory where the json file is generated is excluded by the gitignore.

paulfantom commented:
rebased... again
